### PR TITLE
Warn when torch._C._set_grad_enabled is called

### PR DESCRIPTION
**Pull request by kshitij12345**
Fixes: https://github.com/Lightning-AI/lightning-thunder/issues/332

Currently, on main, when `_set_grad_enabled` is called via `torch.enable_grad()` or `torch.no_grad()`, it is silently ignored. thunder decides whether to compute gradients based on input setting `requires_grad=True`. This PR adds a warning stating the same so that the user knows `torch.{enable_grad/no_grad}` will not have any effect when used with `thunder.jit`.

```python
import torch
import thunder

def foo(x):   
    with torch.no_grad():
        return x + 1

jfoo = thunder.jit(foo)
x = torch.randn(3, requires_grad=True)
o = jfoo(x)
o.sum().backward()
print(x.grad)  # grad exists

# print(thunder.last_traces(jfoo))

# # Constructed by Delete Last Used (took 0 milliseconds)
# import torch
# from thunder.executors.torchex import no_autocast

# @torch.no_grad()
# @no_autocast
# def augmented_forward_fn(x):
#   # x: "cpu f32[3]"
#   t0 = torch.add(x, 1.0)  # t0: "cpu f32[3]"
#     # t0 = ltorch.add(x, 1.0, alpha=None)  # t0: "cpu f32[3]"
#       # t0 = prims.add(x, 1.0)  # t0: "cpu f32[3]"
#   return {'output': t0, 'flat_args': [x], 'flat_output': (t0,)}, ((), ())
```

### CODE REVIEW

File: thunder/torch/__init__.py [hunk 1]
```diff
@@ -4394,7 +4402,14 @@ def torch_device(device_or_str: DeviceLike, /, index: int | None = None) -> devi
 
 # We don't use @torchsymbol as we don't want `torch.device()` to appear in trace as a symbol.
 # Because of this, we need to manually register the implementation.
-_torch_to_thunder_function_map[torch.device] = torch_device
+register_function(torch.device, torch_device)
+
+
+def _set_grad_enabled_with_warning(enabled: bool) -> None:
```
Comment (mruberry) [inline thunder/torch/__init__.py:4408]: A warning seems like a good start

**Review by mruberry — APPROVED** [review]
